protocols/imap/src/main/scala/org/apache/james/imap/decode/parser/SearchCommandParser.scala

```diff
@@ -0,0 +1,683 @@
+    reader.consumeChar('(')
+    reader.nextWordChar
+    var cap = consumeAndCap(reader)
+    while (cap != ')') {
+      cap match {
+        case 'A' =>
+          expectCharacters("all", 1)(reader)
+          options.append(SearchResultOption.ALL)
+        case 'C' =>
+          expectCharacters("count", 1)(reader)
+          options.append(SearchResultOption.COUNT)
+        case 'M' =>
+          consumeAndCap(reader) match {
+            case 'A' =>
+              expectCharacters("max", 2)(reader)
+              options.append(SearchResultOption.MAX)
+            case 'I' =>
+              expectCharacters("min", 2)(reader)
+              options.append(SearchResultOption.MIN)
+            case _ =>
+              throw new DecodingException(HumanReadableText.ILLEGAL_ARGUMENTS, "Unknown search key")
+          }
+
+        // Check for SAVE options which is part of the SEARCHRES extension
+        case 'S' =>
+          expectCharacters("save", 1)(reader)
+          options.append(SearchResultOption.SAVE)
+        case _ =>
+          throw new DecodingException(HumanReadableText.ILLEGAL_ARGUMENTS, "Unknown search key")
+      }
+      reader.nextWordChar
+      cap = consumeAndCap(reader)
+    }
+    // if the options are empty then we parsed RETURN () which is a shortcut for ALL.
+    // See http://www.faqs.org/rfcs/rfc4731.html 3.1
+    if (options.isEmpty) Seq(SearchResultOption.ALL)
+    else options.toSeq
+  }
+
+  @throws[DecodingException]
+  override protected def decode(request: ImapRequestLineReader, tag: Tag, useUids: Boolean, session: ImapSession): ImapMessage = try {
+    var maybeRecent: Option[SearchKey] = None
+    var options = Seq.empty[SearchResultOption]
+    val c = ImapRequestLineReader.cap(request.nextWordChar)
+    if (c == 'R') { // if we found a R its either RECENT or RETURN so consume it
+      request.consume
+      nextIs(request, 'e')
+      consumeAndCap(request) match {
+        case 'C' =>
+          maybeRecent = Some(recent(request, alreadyConsumed = 2))
+        case 'T' =>
+          expectCharacters("return", 2)(request)
+          request.nextWordChar
+          options = parseOptions(request)
+        case _ =>
+          throw new DecodingException(HumanReadableText.ILLEGAL_ARGUMENTS, "Unknown search key")
+      }
+    }
+    val finalKey = maybeRecent.map { recent =>
+      if (request.nextChar != ' ') {
+        request.eol()
+        recent
+      }
+      else { // Parse the search term from the request
+        val key = decode(session, request)
+        And(Seq(recent, key))
+      }
+    }.getOrElse(decode(session, request))
+
+    import scala.jdk.CollectionConverters._
+    new SearchRequest(new SearchOperation(finalKey, options.asJava), useUids, tag)
+  } catch {
+    case e@(_: IllegalCharsetNameException | _: UnsupportedCharsetException) =>
+      SearchCommandParser.LOGGER.debug("Unable to decode request", e)
+      unsupportedCharset(tag)
+  }
+}
```

Comment:
I "thought" our test coverage was too low to alter this class?

Reply:
I assumed the opposite, any evidence about that?

Reply:
https://github.com/linagora/james-project/pull/2997#pullrequestreview-331942425

> The testsuite is not good. I changed 3 random parts of the code and no test broke.
I'm not sure to agree with refactoring such a major piece of code without a serious coverage (that means testing for errors not only nominal cases).

So I'm intereseted of learning what makes the two refactoring different. Or if your opinion evolved.

Reply:
Fair enough. Here is my summary of what was said in https://github.com/linagora/james-project/pull/2997#pullrequestreview-331942425 : 

* I told you it's badly covered by unit tests and that you did changed the behavior of the code
* you told me that MPT is covering much of it
* you proposed a change to recover the size limitation part of the parser back (the behavior you did changed)

What I did is quite different because I didn't broke refactoring rules: I converted from Java to Scala with Intellij feature and then I only did some changes on the shape of the code: the risk of regression is tiny.

However, to be honest, I would understand someone ask for evidences that the test coverage is good enough

Reply:
I know that the work proposed here is less aggressive than the refactoring I proposed.

However https://github.com/linagora/james-project/pull/3309#discussion_r411230649 is IMO non trivial chages that may justify similar concerns to those I faced.

Reply:
If you believe the risk to be tiny then let's risk it ;-)

Reply:
> However #3309 (comment) is IMO non trivial chages that may justify similar concerns to those I faced.

Oh, come on, it's a really easy one. I prefer to revert this part than having to debate if it's sure or not.